### PR TITLE
[Rich Text Editor] Add link plugin and buttons

### PR DESCRIPTION
**Pull request by Feranchz**
### New `withLinks` plugin:
- This links makes all the elements with `type == 'link'` inline nodes.
- Add a new property to the editor object: `lastInsertedLinkPath`
- Intercept the `insert_node` operation when inserting a new link, when that happens we'll update our `lastInsertedLinkPath`
- Intercept other operations as well to keep the path updated (When wrapping a text in a link, slateJS makes different operations, so we need to always update the path)

### Link button on toolbar:
- To click this button you should have some text highlighted in the **same** block. (You can't highlight text from two blocks and create a link from them) 
- When the link button is clicked, we add an empty link with the selection

### Links:
- Each link rendered has its own popover, with a click you can open them. We automatically open a Popover if the path of that link equals to the `lastInsertedLinkPath` on the editor object.
- When you edit the **url** of a link we only edit the link object. But if you change the text of the link in the popover we need to remove all the childrens of the link and replaced them with the new text

### Backend validations: 
- SlateJS has some constraints that we need to follow. One of them is:[ Inline nodes cannot be the first or last child of a parent block, nor can it be next to another inline node in the children array.](https://docs.slatejs.org/concepts/11-normalizing#built-in-constraints). This means that we can end having this structure:
```
{
  type: 'paragraph',
  children: [
    {
      text: ''
    },
    {
      type: 'link',
      children: [{ type: 'text', text: 'Link' }]
    },
    {
      text: ''
    }
  ]
}
```
- so we need to accept empty nodes on our backend validations

## How to test it
- Create a new Content Type with a field with blocks.
- Try to create and save links

### CODE REVIEW

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/useBlocksStore.js [hunk 1]
```diff
@@ -201,6 +219,144 @@ const handleEnterKeyOnList = (editor) => {
...
+  const elementText = element.children.map((child) => child.text).join('');
+
+  const handleOpenEditPopover = (e) => {
+    e.preventDefault();
+    setPopoverOpen(true);
+  };
+
+  const handleSave = (e) => {
+    e.stopPropagation();
+    const { text, url } = e.target.elements;
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/useBlocksStore.js:241]: I think you should be using [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) imo it's a more atypical approach to working with Forms.

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { Path } from 'slate';
+
+const withLinks = (editor) => {
+  const { isInline, apply } = editor;
+
+  editor.isInline = (element) => {
+    return element.type === 'link' ? true : isInline(element);
+  };
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js:8]: Why are you overwriting this function? Can you write comments please

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    } else if (operation.type === 'move_node') {
+      // We need to update the last inserted link path when link is moved
+      // If link is the first word in the paragraph we dont need to update the path
+      if (Path.hasPrevious(operation.path)) {
+        editor.lastInsertedLinkPath = Path.transform(editor.lastInsertedLinkPath, operation);
+      }
+    }
+
+    apply(operation);
+  };
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js:28]: Same here

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/utils/links.js
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/utils/links.js:1]: Can we add JSDOC types to these function please, helps convert to TS in the future 😄

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/useBlocksStore.js [hunk 4]
```diff
@@ -201,6 +219,144 @@ const handleEnterKeyOnList = (editor) => {
...
+        {children}
+      </BaseLink>
+      {popoverOpen && (
+        <Popover
+          source={linkRef}
+          onDismiss={handleDismiss}
+          padding={4}
+          placement="right-end"
+          contentEditable={false}
+          onEscapeKeyDown={() => ReactEditor.focus(editor)}
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/useBlocksStore.js:278]: is `onDismiss` not called when we press Escape? Just wondering why we're doubling down on focus

**Review by joshuaellis — CHANGES_REQUESTED** [review]
Great work @Feranchz, few minor comments – everyone else probably has a much better view of the architecture of the component & how you're testing stuff so take my comments with a pinch of salt 😄

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js [hunk 5]
```diff
@@ -394,6 +401,53 @@ ListButton.propTypes = {
...
+    <ToolbarButton
+      icon={Link}
+      name="link"
+      label={{
+        id: 'components.Blocks.link',
+        defaultMessage: 'Link',
+      }}
+      isActive={isActive}
+      handleClick={addLink}
+      disabled={selectionIsCollapsed || !selectionIsInSameBlock}
```
Comment (madhurisandbhor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:446]: not sure what is this 2nd condition for, but selectionIsCollapsed also checking the same and 1 step more, but you are negating 2nd condition, its confusing

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js [hunk 6]
```diff
@@ -25,14 +26,15 @@ const FlexButton = styled(Flex).attrs({ as: 'button' })`
   }
 `;
 
-const ToolbarButton = ({ icon, name, label, isActive, handleClick }) => {
+const ToolbarButton = ({ icon, name, label, isActive, handleClick, disabled = false }) => {
```
Comment (madhurisandbhor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:29]: disabled = false, you have this in default props as well

File: packages/core/strapi/src/services/entity-validator/blocks-validator.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import { yup } from '@strapi/utils';
 
 const textNodeValidator = yup.object().shape({
-  type: yup.string().equals(['text']).required(),
```
Comment (markkaylor) [inline packages/core/strapi/src/services/entity-validator/blocks-validator.ts:4]: I don't understand, why is it no longer required. It doesn't match the schema we defined in the RFC

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/tests/useBlocksStore.test.js [hunk 8]
```diff
@@ -1,25 +1,35 @@
...
   },
 ];
 
+const user = userEvent.setup();
+
 const baseEditor = createEditor();
 
 const Wrapper = ({ children }) => {
-  const editor = React.useMemo(() => withReact(baseEditor), []);
+  const editor = React.useMemo(() => withReact(withLinks(baseEditor)), []);
```
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/tests/useBlocksStore.test.js:32]: I just noticed we have a useMemo here but in our editor https://github.com/strapi/strapi/blob/e7e82d57df7e80015b9e78fb0ff8ea2d0656d049/packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js#L43C1-L44C1

we do not. Shouldn't we have the same setup in our tests?

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  const { isInline, apply } = editor;
+
+  // Links are inline elements, so we need to override the isInline method for slate
+  editor.isInline = (element) => {
+    return element.type === 'link' ? true : isInline(element);
+  };
+
+  // We keep a track of the last inserted link path
+  // So we can show the popover on the link component if that link is the last one inserted
+  editor.lastInsertedLinkPath = null;
```
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js:13]: Is this a property we are creating or overriding?

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js:1]: Good idea to create a plugins folder 👍

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/utils/links.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ *
+ * @param {import('slate').Editor} editor
+ * @param {object} link
+ * @param {string} link.url
+ */
+const insertLink = (editor, { url }) => {
+  if (editor.selection) {
+    // We want to remove all link on the selection
+    // We use a try catch because Slate throw error when there is no link in the selection
```
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/utils/links.js:37]: I removed the try catch but unable to reproduce the error:
- I select some text not containing a link
- I click the link button
- Add a link
- Hit save, and it works
Maybe I'm missing something. 

I'm worried we could end up silencing an error that would be difficult to track down if we ever actually have a bug

File: packages/core/strapi/src/services/metrics/__tests__/admin-user-hash.test.ts [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import crypto from 'crypto';
+import createContext from 'strapi/test/helpers/create-context';
```
Comment (markkaylor) [inline packages/core/strapi/src/services/metrics/__tests__/admin-user-hash.test.ts:2]: Just curious why this change is here? It's intentional?

File: packages/core/strapi/src/services/entity-validator/blocks-validator.ts [hunk 12]
```diff
@@ -27,14 +27,10 @@ const linkNodeValidator = yup.object().shape({
...
 const inlineNodeValidator: any = yup.lazy((value: { type: string }) => {
   switch (value.type) {
-    case 'text':
-      return textNodeValidator;
     case 'link':
       return linkNodeValidator;
     default:
-      return yup.mixed().test('invalid-type', 'Inline node must be Text or Link', () => {
-        return false;
-      });
```
Comment (markkaylor) [inline packages/core/strapi/src/services/entity-validator/blocks-validator.ts:37]: I'm not really opposed but I'm wondering what your reasoning is for changing this?

**Review by markkaylor — COMMENTED** [review]
Nice work, it's shaping up. This one was not easy.

I'm unable to deselect a link. When I deselect a link the editor loses focus.


https://github.com/strapi/strapi/assets/26598053/9dc0cad3-e760-4144-b9a4-dc32582b5f22

File: packages/core/strapi/src/services/entity-validator/blocks-validator.ts (hunk 7 above)
Comment (Feranchz) [reply packages/core/strapi/src/services/entity-validator/blocks-validator.ts:4]: Slate adds blocks with this structure `{ text: 'hello' }` without type. These are leafs and (in theory) they are not inline nodes.

One of the rules of Slate said: You can't have an inline node as first/last child of a block, so when you have a block which is only a link, or a link at the beginning/end, Slate automatically will add this without type node. (You can read more about this in the PR description).

I know this doesn't match our schema, but the only way I found to avoid slate doing this is replacing completely the normalize function, which is possible but I don't think makes too much sense

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js (hunk 9 above)
Comment (Feranchz) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js:13]: We are creating it

File: packages/core/strapi/src/services/metrics/__tests__/admin-user-hash.test.ts (hunk 11 above)
Comment (Feranchz) [reply packages/core/strapi/src/services/metrics/__tests__/admin-user-hash.test.ts:2]: Not idea how this get here 🤔 I will remove it

File: packages/core/strapi/src/services/entity-validator/blocks-validator.ts (hunk 7 above)
Comment (markkaylor) [reply packages/core/strapi/src/services/entity-validator/blocks-validator.ts:4]: But couldn't we create a new validator specific for text specific to links. That way it's only an exception for links and we can still be strict with the other nodes.

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js (hunk 9 above)
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/plugins/withLinks.js:13]: I'm wondering how we could document that because maybe it will get confusing later on. Someone in the future might look for this in the Slate docs and not understand why it isn't there for example.

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js [hunk 13]
```diff
@@ -394,6 +401,39 @@ ListButton.propTypes = {
   }).isRequired,
 };
 
+const LinkButton = () => {
+  const editor = useSlate();
+
+  const isLinkActive = isBlockActive(editor, (node) => node.type === 'link');
```
Comment (simotae14) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:407]: Hey @Feranchz , in my Pr https://github.com/strapi/strapi/pull/18181 I removed the definition of isBlockActive because I found a different way to solve my issue with mixed blocks selection in the editor, my suggestion is to move the definition of isBlockActive in the component where you need it, find a better name for it and also you can find the matchType definition directly here https://github.com/strapi/strapi/blob/e7e82d57df7e80015b9e78fb0ff8ea2d0656d049/packages/core/admin/admin/src/content-manager/components/BlocksEditor/hooks/useBlocksStore.js#L366 so you can just use the hook to retrieve the matchNode function for the link, thanks

File: packages/core/strapi/src/services/entity-validator/blocks-validator.ts (hunk 7 above)
Comment (Feranchz) [reply packages/core/strapi/src/services/entity-validator/blocks-validator.ts:4]: @markkaylor  Other option is extend the `normalizeNode` from Slate to add `type: text` to all the text nodes. If you see [here](https://docs.slatejs.org/api/nodes/text) for them a Text node don't have the need to have a type field, we are adding it to match our schema, so makes sense we make sure we always have a type by extending the function

Comment (markkaylor) [reply packages/core/strapi/src/services/entity-validator/blocks-validator.ts:4]: Nice! Yeah that sounds like a good idea.

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js [hunk 14]
```diff
@@ -58,7 +59,9 @@ const withImages = (editor) => {
 const BlocksEditor = React.forwardRef(
   ({ intlLabel, labelAction, name, disabled, required, error, value, onChange }, ref) => {
     const { formatMessage } = useIntl();
-    const [editor] = React.useState(() => withReact(withImages(withHistory(createEditor()))));
+    const [editor] = React.useState(() =>
+      withReact(withStrapiSchema(withLinks(withImages(withHistory(createEditor())))))
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js:63]: at what point do we turn this into a pipe fn 😂

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js (hunk 13 above)
Comment (simotae14) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:451]: hey @Feranchz my PR is merged now

**Review by markkaylor — APPROVED** [review]
Nice work Fernando 🎈 that was a tricky one!

**Review by joshuaellis — APPROVED** [review]
Great work @Feranchz, perseverance really paid off with this one 👏🏼
